lib/Dialect/HW/HWOps.cpp

```diff
@@ -1978,21 +2072,51 @@ void StructInjectOp::print(OpAsmPrinter &printer) {
 
 LogicalResult StructInjectOp::canonicalize(StructInjectOp op,
                                            PatternRewriter &rewriter) {
-  auto inputOp = op.getInput().getDefiningOp();
-
-  // b = inject(inject(x["field"], v0)["field"], v1) => inject(x["field"], v1)
-  if (auto structInject = dyn_cast_or_null<StructInjectOp>(inputOp)) {
-    if (structInject == op)
+  // Canonicalize multiple injects into a create op and eliminate overwrites.
+  SmallPtrSet<Operation *, 4> injects;
+  DenseMap<StringAttr, Value> fields;
+
+  // Chase a chain of injects. Bail out if cycles are present.
+  StructInjectOp inject = op;
+  Value input;
+  do {
+    if (!injects.insert(inject).second)
       return failure();
 
-    if (structInject.getField() == op.getField()) {
-      rewriter.replaceOpWithNewOp<StructInjectOp>(
-          op, op.getType(), structInject.getInput(), op.getField(),
-          op.getNewValue());
-      return success();
+    fields.try_emplace(inject.getFieldAttr(), inject.getNewValue());
+    input = inject.getInput();
+    inject = dyn_cast_or_null<StructInjectOp>(input.getDefiningOp());
+  } while (inject);
+  assert(input && "missing input to inject chain");
+
+  auto ty = hw::type_cast<StructType>(op.getType());
+  auto elements = ty.getElements();
+
+  // If the inject chain sets all fields, canonicalize to create.
+  if (fields.size() == elements.size()) {
+    SmallVector<Value> createFields;
+    for (const auto &field : elements) {
+      auto it = fields.find(field.name);
+      assert(it != fields.end() && "missing field");
+      createFields.push_back(it->second);
     }
+    rewriter.replaceOpWithNewOp<StructCreateOp>(op, ty, createFields);
   }
 
+  // Nothing to canonicalize, only the original inject in the chain.
+  if (injects.size() == fields.size())
+    return failure();
+
+  // Eliminate overwrites. The hash map contains the last write to each field.
+  for (const auto &field : elements) {
+    auto it = fields.find(field.name);
+    if (it == fields.end())
+      continue;
+    input = rewriter.create<StructInjectOp>(op.getLoc(), ty, input, field.name,
+                                            it->second);
+  }
```

Comment:
I think alternatively we can explicitly create a struct with injected elements and original inputs. I guess you want to create chains of struct_inject as preparation for PrettifyVerilog but for use cases other than register assignments, it looks more readable if we can create struct_create directly, e.g:

```
// s: struct<a, b, c>
%1 = inject(%s["a"], %v1)
%2 = inject(%1["b"], %v2)
%3 = inject(%2["a"], %v3)
```
==>
```
%0 = create(a:%v3, b: %v2, c: %s["c"]) 
```

WDYT?

Reply:
I wouldn't like to make this decision now. One thing we could also consider to improve this is to make the inject opt variadic.